### PR TITLE
WIP: Top sort

### PR DESCRIPTION
**Pull request by phadej**
Uses `algebraic-graphs`, which will support GHC-7.8 soon. https://github.com/snowleopard/alga/pull/28 Is it ok to drop GHC-7.6 support from cabal-plan?

topo2

### CODE REVIEW

File: src-exe/cabal-plan.hs [hunk 1]
```diff
@@ -106,18 +134,97 @@ doInfo (plan,projbase) = do
...
                 let Just dep' = M.lookup dep pm
                     pid = uPId dep'
                 putStrLn ("  " ++ T.unpack (dispPkgId pid))
             putStrLn ""
 
     return ()
 
+doDot :: Bool -> Bool -> PlanJson -> IO ()
+doDot showBuiltin showGlobal plan = do
+    putStrLn "digraph plan {"
```
Comment (snowleopard) [inline src-exe/cabal-plan.hs]: Alga has built-in `Algebra.Graph.Export.Dot` which you might find useful:

http://hackage.haskell.org/package/algebraic-graphs-0.0.5/docs/Algebra-Graph-Export-Dot.html#v:export

**Comment by hvr** [comment]
Although `alga` looks well-made, we already got a topological sorting function implemented in `src-exe/cabal-plan.hs`, see `toposort :: Ord a => Map a (Set a) -> [a]`; I haven't yet looked at the patch in detail, but wouldn't that fit the bill?

File: src-exe/cabal-plan.hs (hunk 1 above)
Comment (phadej) [reply src-exe/cabal-plan.hs]: Yes, I'd used that, if not the version restrictions on *alga* atm. (That code is part of already merged PR).

**Comment by phadej** [comment]
@hvr, as I mentioned `alga` will support GHC-7.8 *soon*. How strict (and why, out of curiosity) you need to support GHC-7.6. That's possible too, but I was lazy to go thru GHC trac (there's one problem in the test-suite, I'd like to know if there exists some workaround).

**Comment by phadej** [comment]
And why `alga`: When filtering the graph, I'd need to filter the "neighbors". In fact, I don't know what happens if in `AdjacencyMap` there are neighbors which doesn't exist as a key of map.

TL;DR some proper graph library would be nice, and `Data.Graph` isn't proper.

File: cabal-plan.cabal [hunk 2]
```diff
@@ -62,6 +62,7 @@ executable cabal-plan
     -- dependencies w/ inherited version ranges via 'cabal-plan' library
     build-depends: cabal-plan
                  , base
+                 , algebraic-graphs  == 0.0.5
```
Comment (hvr) [inline cabal-plan.cabal]: 1. Why the odd version constraint? Does `algebraic-graphs` fail to follow the PVP?
2. as the comment stats, this group of build-depends is for those whose constraints are inherited from the library part; there's a second group below for other deps.

**Comment by hvr** [comment]
@phadej ok, I missed the comment about ghc-7.8; I guess I can stomach losing GHC 7.6 support; besides having support for at least GHC 7.8 I'd also require dependencies to fully comply with the PVP; If `algebraic-datastructures` fits that bill, I'm down with it.

File: cabal-plan.cabal (hunk 2 above)
Comment (snowleopard) [reply cabal-plan.cabal]: >  Does `algebraic-graphs` fail to follow the PVP?

@hvr From the library description: `This is an experimental library and the API will be unstable until version 1.0.0`. However, it looks like I have to reconsider this given that Alga already has several users. So I can start following the PVP from the next release if need be.

**Comment by snowleopard** [comment]
> In fact, I don't know what happens if in `AdjacencyMap` there are neighbors which doesn't exist as keys of the map.

@phadej You cannot construct an `AdjacencyMap` with an edge referring to a non-existing vertex. This consistency invariant is maintained by the public API, however one can violate it via the `Internal` module.

> `alga` could use a `topSort` variant which doesn't abort if there are cycles, btw.

What should this `topSort` variant return if there are cycles in the graph? It seems that this is a job for the `scc` function which handles cycles by combining them into strongly connected components. Feel free to raise an issue to discuss.

File: src-exe/cabal-plan.hs [hunk 3]
```diff
@@ -182,6 +190,42 @@ doDot showBuiltin showGlobal plan = do
...
+            | unit <- toList units
+            , depId <- toList $ foldMap (\ci -> ciLibDeps ci <> ciExeDeps ci) (uComps unit)
+            , depId /= uId unit
+            , dep <- maybeToList $ M.lookup depId units
+            ]
+
+        am :: AM.AdjacencyMap UnitId
+        am = AlgaC.toGraph (fmap uId g)
+
+    for_ (AM.topSort am) $ \xs -> for_ xs $ \unitId ->
```
Comment (snowleopard) [inline src-exe/cabal-plan.hs]: It looks like you wouldn't need to use `AlgaC` or `AM` if `Algebra.Graph` exported `topSort` on the `Graph` datatype. This is easy to implement and would make the code here much nicer.

File: cabal-plan.cabal (hunk 2 above)
Comment (hvr) [reply cabal-plan.cabal]: @snowleopard Well, being experimental and following the PVP is not a contradicition (it just means you'll have to make sure to major bump more frequently; long time ago, we had a stabilify-flag in the cabal file for that, but it was considered pointless. When uploading to Hackage, we expect you to follow the PVP; if your package is too experimental for other packages start  depending upon it, it doesn't belong on Hackage yet; see also http://hackage.haskell.org/upload ; you touched a topic that's  quite a pet peeve of mine as it's likely the one area of the Haskell ecosystem I've invested most time and effort in... :-)

Comment (snowleopard) [reply cabal-plan.cabal]: @hvr Agreed. The next release version will be `0.1.0` and Alga will comply with PVP.

File: src-exe/cabal-plan.hs (hunk 3 above)
Comment (phadej) [reply src-exe/cabal-plan.hs]: can there be `topSort` on `Graph a`, without `Ord a`, that would be great.

**Comment by phadej** [comment]
@snowleopard about topological sorting and cyclic graphs, 
- http://lkozma.net/blog/ordering-cyclic-graphs/

> What if the graph has cycles (two teams that played a draw, or a sequence of teams where A beat B, B beat C, …, X beat A), but we still want to output a reasonable ordering? What is a reasonable ordering?

> This is the vertex ordering problem and we can formulate many different criteria. ... a reasonable criterion should actually lead to the topological sorting when there are no cycles.

E.g. [Minimum feedback arc set](https://en.wikipedia.org/wiki/Feedback_arc_set) or [minimum bandwidth](https://en.wikipedia.org/wiki/Graph_bandwidth). FWIW, I do not know what `topSort` from `containers` does (e.g. with a `A->B->C->D->A + A->C` graph, will `A` be starting point?)

File: src-exe/cabal-plan.hs (hunk 3 above)
Comment (snowleopard) [reply src-exe/cabal-plan.hs]: Hmm, interesting. `Eq a` is definitely required. Without `Ord a` I'm afraid `topSort` will become quite expensive, because we essentially will have to implement the `Eq a`-based `nub` functionality: given a graph expression `a + d + a + c + b + a + b` we'll need to produce the list `[a, d, c, b]`.

In short: yes, we can, but it will be too slow to be useful.

**Comment by snowleopard** [comment]
@phadej Both minimum feedback arc set and minimum bandwidth problems are NP-hard, so will likely require some approximate solutions in practice. Feel free to open an issue, but it may take while to figure out a good functional implementation.
